rust/datafusion/src/execution/expression.rs

```diff
@@ -334,44 +337,52 @@ pub fn compile_scalar_expr(
                 Ok(RuntimeExpr::Compiled {
                     name: col.name().clone(),
                     t: col.data_type().clone(),
-                    f: match col.data_type() {
-                        DataType::Int8 => {
-                            cast_column_outer!(index, Int8Array, &data_type)
-                        }
-                        DataType::Int16 => {
-                            cast_column_outer!(index, Int16Array, &data_type)
-                        }
-                        DataType::Int32 => {
-                            cast_column_outer!(index, Int32Array, &data_type)
-                        }
-                        DataType::Int64 => {
-                            cast_column_outer!(index, Int64Array, &data_type)
-                        }
-                        DataType::UInt8 => {
-                            cast_column_outer!(index, UInt8Array, &data_type)
-                        }
-                        DataType::UInt16 => {
-                            cast_column_outer!(index, UInt16Array, &data_type)
-                        }
-                        DataType::UInt32 => {
-                            cast_column_outer!(index, UInt32Array, &data_type)
-                        }
-                        DataType::UInt64 => {
-                            cast_column_outer!(index, UInt64Array, &data_type)
-                        }
-                        DataType::Float32 => {
-                            cast_column_outer!(index, Float32Array, &data_type)
-                        }
-                        DataType::Float64 => {
-                            cast_column_outer!(index, Float64Array, &data_type)
-                        }
-                        _ => panic!("unsupported CAST operation"), /*TODO */
-                                                                   /*Err(ExecutionError::NotImplemented(format!(
-                                                                       "CAST column from {:?} to {:?}",
-                                                                       col.data_type(),
-                                                                       data_type
-                                                                   )))*/
-                    },
+                    f: Rc::new(|batch: &RecordBatch| {
+                        // match compute::cast(batch.column(index), data_type) {
+                        //     Ok(array) => Ok(array),
+                        //     Err(e) => Err(e.into())
+                        // }
+                        compute::cast(batch.column(index).clone(), data_type.clone())
```

Comment:
@andygrove this is where I'm getting the error

Reply:
I can help with this. I'll take a look now and see if it is easy to fix or not.

Reply:
Here you go. This works:

```rust
&Expr::Column(index) => {
                let col = input_schema.field(index);
                let dt = data_type.clone();
                Ok(RuntimeExpr::Compiled {
                    name: col.name().clone(),
                    t: col.data_type().clone(),
                    f: Rc::new(move |batch: &RecordBatch| {
                        compute::cast(batch.column(index), &dt)
                            .map_err(|e| ExecutionError::ArrowError(e))
                    }),
```

The issue was that the closure being built had a reference to data_type from the expression. The fix was to clone it before creating the closure, and use the `move` keyword to move ownership of the variables into the closure.

Reply:
Awesome! I would have never figured. Code's compiling, so I should be able to complete this PR by the weekend